### PR TITLE
added trace events for scheduling responses to platform messages

### PR DESCRIPTION
**Pull request by gaaclarke**
This measures the latency between thread hops when sending messages from Flutter->Host and responding Host->Flutter.

We have platform channel channel profiling on the framework side, but it can't determine how much of the round trip time is just the latency between scheduling on a thread and it actually getting serviced.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: lib/ui/window/platform_message_response_dart.cc [hunk 1]
```diff
@@ -8,17 +8,52 @@
...
+                    bool* is_complete,
+                    const std::string& channel,
+                    Result&& result) {
+  if (callback.is_empty()) {
+    return;
+  }
+  FML_DCHECK(!*is_complete);
+  *is_complete = true;
+  uint64_t platform_message_id = platform_message_counter.fetch_add(1);
+  TRACE_EVENT_ASYNC_BEGIN1("flutter", "PlatformChannel:ScheduleResult",
```
Comment (dnfield) [inline lib/ui/window/platform_message_response_dart.cc:32]:
```suggestion
  TRACE_EVENT_ASYNC_BEGIN1("flutter", "PlatformChannel::ScheduleResult",
```

File: lib/ui/window/platform_message_response_dart.cc [hunk 2]
```diff
@@ -8,17 +8,52 @@
...
+                           platform_message_id, "channel", channel.c_str());
+  ui_task_runner->PostTask(fml::MakeCopyable(
```
Comment (dnfield) [inline lib/ui/window/platform_message_response_dart.cc:34]: We seem to have dropped the idea of callback being empty here?

Comment (gaaclarke) [reply lib/ui/window/platform_message_response_dart.cc:34]: Nope, it's on line 26 which matches the previous logic.

File: lib/ui/window/platform_message_response_dart.cc [hunk 3]
```diff
@@ -28,42 +63,16 @@ PlatformMessageResponseDart::~PlatformMessageResponseDart() {
 }
 
 void PlatformMessageResponseDart::Complete(std::unique_ptr<fml::Mapping> data) {
-  if (callback_.is_empty()) {
-    return;
-  }
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply lib/ui/window/platform_message_response_dart.cc:33]: Line 26.

File: shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java [hunk 4]
```diff
@@ -312,8 +312,10 @@ private void dispatchMessageToQueue(
       int replyId,
       long messageData) {
     final DartMessengerTaskQueue taskQueue = (handlerInfo != null) ? handlerInfo.taskQueue : null;
+    TraceSection.beginAsyncSection("PlatformChannel:ScheduleHandler", replyId);
```
[the remark this answers is not in the document]
Comment (gaaclarke) [reply shell/platform/android/io/flutter/embedding/engine/dart/DartMessenger.java:315]: @jiahaog done.

**Comment by jmagman** [comment]
Friendly bump, is this waiting on another review from @dnfield?

**Review by dnfield — APPROVED** [review]

**Pull request merged**